### PR TITLE
Cprod 682

### PR DESCRIPTION
**Pull request by hagsteel**
* Adding versioned storage to the canister derive macro

### CODE REVIEW

File: ic-canister/ic-canister-macros/src/derive.rs [hunk 1]
```diff
@@ -213,28 +201,38 @@ fn expand_upgrade_methods(
...
+                ::ic_storage::stable::write(#state_borrow).unwrap();
             }
 
             #[cfg(all(target_arch = "wasm32", feature = "export_api"))]
             #[export_name = "canister_post_upgrade"]
             fn __post_upgrade() {
                 use ::ic_storage::IcStorage;
 
-                let (#( #field_names,)*) = ::ic_cdk::storage::stable_restore().unwrap();
+                let #name = match ::ic_storage::stable::read::<#field_type::Versioned>() {
```
Comment (Maximkaaa) [inline ic-canister/ic-canister-macros/src/derive.rs:214]: Shouldn't this be `::Previous`?

File: ic-canister/ic-canister-macros/src/derive.rs [hunk 2]
```diff
@@ -213,28 +201,38 @@ fn expand_upgrade_methods(
...
+                    Ok(val) => val,
+                    Err(e) => panic!("failed to upgrade: {}", e),
```
Comment (Maximkaaa) [inline ic-canister/ic-canister-macros/src/derive.rs:216]: For debugging purposes, I think it's better to use `::ic_cdk::trap` here

File: ic-canister/ic-canister-macros/src/derive.rs [hunk 3]
```diff
@@ -213,28 +201,38 @@ fn expand_upgrade_methods(
...
 
-                #( #fields_assignment )*
+                #field_assignment
             }
         }
     }
 }
 
 fn is_state_field_stable(_field: &Field) -> bool {
+    // path = stable_store
```
Comment (Maximkaaa) [inline ic-canister/ic-canister-macros/src/derive.rs:226]: Do we need this commented stuf?

File: ic-canister/ic-canister-macros/src/derive.rs [hunk 4]
```diff
@@ -103,8 +103,8 @@ pub fn derive_canister(input: TokenStream) -> TokenStream {
                 .push(quote! {#field_name : <#field_type as ::ic_storage::IcStorage>::get()});
             state_fields_test.push(quote! {#field_name : ::std::rc::Rc::new(::std::cell::RefCell::new(<#field_type as ::std::default::Default>::default()))});
 
-            if is_stable {
-                stable_fields.push((field_name, field_type));
+            if is_stable && stable_field.is_none() {
```
Comment (Maximkaaa) [inline ic-canister/ic-canister-macros/src/derive.rs:106]: Doesn't ths mean that if the `State` struct has more then one stable field, the macro will just disregard all except the first one. I think we want to throw a compile error in this case.

File: ic-canister/ic-canister-macros/src/derive.rs (hunk 1 above)
Comment (hagsteel) [reply ic-canister/ic-canister-macros/src/derive.rs]: Good spot!
Changed and updated

File: ic-canister/ic-canister-macros/src/derive.rs (hunk 2 above)
Comment (hagsteel) [reply ic-canister/ic-canister-macros/src/derive.rs]: Good point

File: ic-canister/ic-canister-macros/src/derive.rs (hunk 4 above)
Comment (hagsteel) [reply ic-canister/ic-canister-macros/src/derive.rs]: Fixed

File: ic-canister/ic-canister-macros/src/derive.rs (hunk 3 above)
Comment (hagsteel) [reply ic-canister/ic-canister-macros/src/derive.rs]: Nope, implemented the function instead.

**Review by blutooth — APPROVED** [review]

**Review by blutooth — APPROVED** [review]

**Pull request merged**